### PR TITLE
feat : Replace tsnecuda with openTSNE for embeddings t-SNE

### LINKED ISSUE #1501: T-SNE embeddings only support CPU backend (bhtsne), no GPU acceleration
## Description

  `tsnecuda` has been unmaintained since 2022 and requires NVIDIA GPU + CUDA to
  run.
  it should be replaced with `openTSNE` which is:
  - Actively maintained
  - CPU-based (no GPU/CUDA requirement)
  - pip-installable (`pip install openTSNE`)
  - Fast enough for visdom's embedding visualization use case

---

  ## Current Behavior
  Visdom uses `tsnecuda` as the primary t-SNE backend for `vis.embeddings()`,
  falling back to `bhtsne`.

---

  ## Expected Behavior
  Use `openTSNE` as the primary backend, with `bhtsne` as fallback.

### PR DESCRIPTION
**Pull request by Manik-Khajuria-5**
## Description

  Replace unmaintained `tsnecuda` (GPU) with actively maintained `openTSNE` (CPU) as the primary t-SNE backend for embeddings visualization. Falls back gracefully to `bhtsne` if unavailable.

---

  ## Motivation and Context

  `tsnecuda` has been unmaintained since 2022 and requires NVIDIA GPU + CUDA. `openTSNE` is actively maintained, pip-installable, CPU-based, and fast enough for visdom's typical embedding sizes.

  ## Fixes: #1501 

---

  ## How Has This Been Tested?

  - Manual UI test with synthetic clustered data (200 points, 5 clusters)  embeddings render correctly with hover, lasso-select, and zoom/pan
  - Black formatting passes

---

  ## Changes
  - `py/visdom/__init__.py`: `from openTSNE import TSNE` replaces `from tsnecuda
   import TSNE`, uses `.fit()` instead of `.fit_transform()`
  - `README.md`: Updated embeddings docs to reference openTSNE

---

  ## Types of changes
  - [x] New feature (non-breaking change which adds functionality)
  - [x] Code refactor or cleanup (changes to existing code for improved
  readability or performance)

---

  ## Checklist:
  - [x] My code follows the code style of this project.

### CODE REVIEW

**Review by Jayantparashar10 — CHANGES_REQUESTED** [review]
@Manik-Khajuria-5 as we discussed in the meeting, tsnecuda was last updated in 2022, so we will need to look for an alternative after the maintenance release.
What do you guys think? @vedansh-5 @rajnisht7 @tonypzy

**Comment by Manik-Khajuria-5** [comment]
Thanks @Jayantparashar10, good point  tsnecuda is unmaintained since 2022. I can replace it with openTSNE which is actively maintained, pip-installable, and has a similar API. Some other options are scikit-learn TSNE, RAPIDS cuML, and FIt-SNE but openTSNE seems like the best fit. Happy to update the PR But tsnecuda is also fine

**Comment by Manik-Khajuria-5** [comment]
@Jayantparashar10 Please reply so I can work according to it

**Comment by Jayantparashar10** [comment]
@Manik-Khajuria-5 I think this is a todo-completion PR. Please remove the "release blocker" label.

**Comment by Jayantparashar10** [comment]
> Thanks @Jayantparashar10, good point tsnecuda is unmaintained since 2022. I can replace it with openTSNE which is actively maintained, pip-installable, and has a similar API. Some other options are scikit-learn TSNE, RAPIDS cuML, and FIt-SNE but openTSNE seems like the best fit. Happy to update the PR But tsnecuda is also fine

I think openTSNE is the best fit because it doesn't require CUDA, has a similar API, and is faster compared to scikit-learnTSNE and other implementations.

**Review by tonypzy — COMMENTED** [review]
Actually, openTSNE is CPU based, not GPU based.

**Comment by tonypzy** [comment]
if so, title and description of PR need to be changed

**Comment by Manik-Khajuria-5** [comment]
@Jayantparashar10 @tonypzy I have implemented openTsne can u re review it please

File: py/visdom/__init__.py [hunk 1]
```diff
@@ -44,43 +44,61 @@
...
 assert sys.version_info[0] >= 3, "To use visdom with python 2, downgrade to v0.1.8.9"
 
+
+def _normalize_tsne(Y):
+    xmin, xmax = min(Y[:, 0]), max(Y[:, 0])
+    ymin, ymax = min(Y[:, 1]), max(Y[:, 1])
+    xrange = xmax - xmin
+    yrange = ymax - ymin
+    normx = ((Y[:, 0] - xmin) / xrange) * 2 - 1 if xrange > 0 else Y[:, 0]
+    normy = ((Y[:, 1] - ymin) / yrange) * 2 - 1 if yrange > 0 else Y[:, 1]
```
Comment (rajnisht7) [inline py/visdom/__init__.py]: @Manik-Khajuria-5 check the normalisation logic for normx and normy, when xrange and yrange ==0

**Review by rajnisht7 — CHANGES_REQUESTED** [review]

File: py/visdom/__init__.py [hunk 2]
```diff
@@ -44,43 +44,61 @@
...
+    return list(zip(normx, normy))
```
[the remark this answers is not in the document]
Comment (Manik-Khajuria-5) [reply py/visdom/__init__.py:61]: resolved

File: py/visdom/__init__.py (hunk 1 above)
Comment (Manik-Khajuria-5) [reply py/visdom/__init__.py]: done

File: py/visdom/__init__.py [hunk 3]
```diff
@@ -44,43 +44,61 @@
...
+
+
+def _get_perplexity(num_entities):
+    if num_entities >= 150:
+        return 50
+    elif num_entities >= 21:
+        return num_entities // 3
+    else:
+        return 7
```
[the remark this answers is not in the document]
Comment (Manik-Khajuria-5) [reply py/visdom/__init__.py]: done

**Review by vedansh-5 — APPROVED** [review]
with the ai comments addressed this PR looks good to me.

**Review by Jayantparashar10 — CHANGES_REQUESTED** [review]
Add openTSNE to requirements.txt for easy install

**Comment by tonypzy** [comment]
> Add openTSNE to requirements.txt for easy install

Vi s do m has no `equirements.txt`.

**Review by tonypzy — COMMENTED** [review]
please resolve reviews

**Comment by Manik-Khajuria-5** [comment]
@tonypzy I have resolved the conflicts please review again

**Review by tonypzy — COMMENTED** [review]
One last thing: openTSNE should be added to set.py so that new user will not meet error when they call vis.embeddings()

**Comment by Manik-Khajuria-5** [comment]
@tonypzy Sure

**Comment by Manik-Khajuria-5** [comment]
@tonypzy Please review again I have made changes

**Review by tonypzy — APPROVED** [review]
LGTM

**Review by Jayantparashar10 — APPROVED** [review]
LGTM

**Review by rajnisht7 — APPROVED** [review]

**Review by Saksham-Sirohi — APPROVED** [review]
